python/lsst/cp/pipe/ptc.py

```diff
@@ -432,85 +485,141 @@ def funcAstier(self, pars, x):
         a00, gain, noise = pars
         return 0.5/(a00*gain*gain)*(np.exp(2*a00*x*gain)-1) + noise/(gain*gain)
 
-    def fitPtcAndNl(self, fitVectorsDict, ptcFitType='POLYNOMIAL'):
-        """Function to fit PTC, and calculate linearity and linearity residual
+    @staticmethod
+    def _initialParsForPolynomial(order):
+        assert(order >= 2)
+        pars = np.zeros(order, dtype=np.float)
+        pars[0] = 10
+        pars[1] = 1
+        pars[2:] = 0.0001
+        return pars
+
+    @staticmethod
+    def _boundsForPolynomial(initialPars):
+        lowers = [np.NINF for p in initialPars]
+        uppers = [np.inf for p in initialPars]
+        lowers[1] = 0  # no negative gains
+        return (lowers, uppers)
+
+    @staticmethod
+    def _boundsForAstier(initialPars):
+        lowers = [np.NINF for p in initialPars]
+        uppers = [np.inf for p in initialPars]
+        return (lowers, uppers)
+
+    @staticmethod
+    def _getInitialGoodPoints(means, variances, maxDeviation):
+        """Return a boolean array to mask bad points.
+
+        A linear function has a constant ratio, so find the median
+        value of the ratios, and exclude the points that deviate
+        from that by more than a factor of maxDeviation.
+
+        Too high and points that are so bad that fit will fail will be included
+        Too low and the non-linear points will be excluded, biasing the NL fit."""
+        ratios = [b/a for (a, b) in zip(means, variances)]
```

Comment:
Should you guard against division by zero? (with an error message or something)

Reply:
Hmmm... that would require the variance to be exactly zero. For that to be the the case, all the data must be bitwise identical, as this is the empirical variance, not the mean of a  (potentially non-existent) variance plane. I think if that's the case, this is a good a place as any for the user to find out that their data is extremely messed up. I think that adding a  check to values _before_ this function isn't a bad idea, but that should be work to add sanity checks to this whole task, and probably belongs on a new tickets (or be added in general when tasks are written).